Extend tracing with Common Trace Format v2

fnde-ot opened the pull request:
Here is the reworked version of PR #8682.

It is rebased on top of latest master and addresses most of the comments from the previous PR. The initial intent was to output a trace similar to that of LLTng, but this should better be done by post-processing scripts if necessary.

Detailed changes:

General:
- rebase on top of latest master
- map all existing trace functions to CTF events
- minor CMake changes to address comments on previous PR
- minor documentation changes to address comments on previous PR

Native POSIX bottom layer:
- select CONFIG_CTF_BOTTOM_POSIX by default for native_posix board when CONFIG_CTF is selected
- add command line option to allow setting the output file
- add an external text-based TSDL metadata file to be used with the CTF output from Zephyr
- add a how-to section in documentation to visualize the output trace with babeltrace/TraceCompass

The only comment not addressed is @pizi-nordic's comment about using __builtin_memcpy() instead of memcpy() in the posix bottom layer. The compiler should be able to optimize if necessary.


Edit in latest push:
- Move documentation to doc/subsystem/tracing
- Move source and headers to subsys/debug except for MIT-licensed file
- Add an extra commit for MIT-licensed file (ctf_map.h) that stays in ext/debug
- Address most documentation comments from @aescolar and verify proper RST formatting
- Change Kconfig options to CONFIG_TRACING_CTF* instead of CONFIG_CTF* to align with CONFIG_TRACING_CPU_STATS
- Use designated initializers for command line argument in ctf_bottom.c
- Use posix_print_error_and_exit instead of __ASSERT in ctf_bottom.c
- Add C++ guards in tracing_ctf.h
- Change CTF_BOTTOM_LOCK and CTF_BOTTOM_UNLOCK to function-like macros
- Rename metadata_timestamped file to metadata

Edit:
- Modify header inclusion guards as per convention (@aescolar)
- Fix TSDL path in documentation
- Use a define and comment on the use of 20 bytes bounded strings.

Edit:
- Address comments from @MarekPieta and @nashif

carlescufi commented:
CC @MarekPieta

dbkinder commented on ext/debug/ctf/README.rst:
```diff
@@ -0,0 +1,216 @@
+.. _ctf:
```
There are also a few reST formatting issues I could comment on, but since we don't include documentation in the ``ext`` folder in our published documentation it's won't show up as an error. This README will be here in the Git repo for folks to read if needed.

fnde-ot replied:
I ran rstcheck and fixed a couple of issues.

dbkinder commented on ext/debug/ctf/README.rst:
```diff
@@ -0,0 +1,216 @@
+CTF middle-layer example
+------------------------
+
+The CTF_EVENT macro will serialize each argument to a field:
```
If you end this line with a double colon (``::``) the following block of indented text will be treated as a code-block (which is probably what you intended).

fnde-ot replied:
Fixed, thanks!

pizi-nordic commented on ext/debug/ctf/bottoms/posix/ctf_bottom.h:
```diff
@@ -0,0 +1,78 @@
+/* No need for locking when ctf_bottom_emit does POSIX fwrite(3) which is thread safe.
+ * Used by middle-layer.
+ */
+#define CTF_BOTTOM_LOCK         { /* empty */ }
```
Could we use function-like invocation of macros that could execute code?
`CFT_BOTTOM_LOCK(); CTF_BOTTOM_UNLOCK();`

fnde-ot replied:
Absolutely! If it makes it more readable then I'm all for it.

nashif commented:
what parts of the PR are actually external code? Looks like we are putting almost everything in ext/ when it shouldn't. The only thing to go into ext/ are the external components, any zephyr related implementation or code needs to be in subsys/debug.

pizi-nordic commented on subsys/debug/tracing/ctf/ctf_top.c:
```diff
@@ -0,0 +1,177 @@
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
```
As far as I remember, this is third copy of `is_idle_thread()` :). Could we make public kernel API from this call?

fnde-ot replied:
I can do this in this same PR (as a separate commit) if you think it makes sense.

fnde-ot replied:
Will be done on a later PR.

pizi-nordic commented on ext/debug/ctf/ctf_middle.h:
```diff
@@ -0,0 +1,215 @@
+
+
+typedef struct {
+	char buf[20];
```
Why 20?

fnde-ot replied:
It seemed like a reasonable upper limit (4x32 bit words) for a string sent out to a potentially bandwidth limited backend. In general we should probably avoid sending strings as available bandwidth can be better used for other things.

fnde-ot replied:
Latest push removes the magic number in favor of a proper define with comment explaining the reasoning.

aescolar reviewed (commented):
Mostly a lot of minors

fnde-ot commented:
>           what parts of the PR are actually external code? Looks like we are putting almost everything in ext/ when it shouldn't. The only thing to go into ext/ are the external components, any zephyr related implementation or code needs to be in subsys/debug.

The bottom layer for POSIX uses some MIT licensed code (ctf_map.h), I can try to move everything except for this file to subsys/debug.

fnde-ot commented on ext/debug/ctf/bottoms/posix/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+zephyr_include_directories(.)
```
If MIT can be relicensed to Apache2, I might as well place everything under subsys/debug. I'll wait for a decision on this.

fnde-ot commented on subsys/debug/tracing/include/tracing_ctf.h:
```diff
@@ -0,0 +1,30 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _TRACE_CTF_H
```
Will do!

fnde-ot commented on ext/debug/ctf/bottoms/posix/ctf_bottom.c:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	ctf_bottom.ostream = fopen(ctf_bottom.pathname, "wb");
+	__ASSERT(ctf_bottom.ostream != NULL,
```
I agree!

fnde-ot committed:
ext: debug: Add C Pre-Processor MAP helper functions for CTF

To be used in the implementation of a POSIX bottom layer to CTF Tracing.

Origin: William Swanson
License: MIT
URL: https://github.com/swansontec/map-macro
commit: 383c38e95e94d4dbd67ec7e31eff80cd317e3f8d
Purpose: Helper macros for CTF POSIX bottom layer.
Maintained-by: External

Signed-off-by: François Delawarde <fnde@oticon.com>

fnde-ot committed:
debug: CTF Tracing with POSIX backend

This commit implements a CTF-backend for Zephyr's tracing API.
The CTF-backend itself is split in a middle-layer and a bottom-layer.
- Middle-layer decides the payload in event transactions,
- Bottom-layer implements the IO transport.

A simple POSIX bottom-layer is provided so far.

Signed-off-by: François Delawarde <fnde@oticon.com>

aescolar commented:
@dbkinder : Could you give a look to the updated documentation?

aescolar commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+
+- Create a new empty directory and copy into it:
+
+  - The TSDL file (``subsys/debug/ctf/tsdl/metadata``)
```
subsys/debug/*tracing/*ctf/tsdl/metadata

aescolar reviewed (approved):
Just 2 minors left, otherwise I'm happy

MarekPieta commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+##############################
+
+Common Trace Format, CTF, is an open format and language to describe trace
+formats.  This enables tool reuse, of which line-textual (babeltrace) and
```
double space before "This"

MarekPieta commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+graphical (TraceCompass) variants already exist.
+
+CTF should look familiar to C programmers but adds stronger typing.
+See `CTF - A  Flexible, High-performance Binary Trace Format
```
double space before "Flexible"

MarekPieta replied:
double space before "Flexible"?

MarekPieta commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+
+  - asynchronous:
+                  means data-transmission is pending or ongoing with the return
+                  of the call. Usually, interrupts/callbacks/signals or polling
```
"polling" written twice

MarekPieta commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+
+
+Unfortunately, I/O may not be atomic and may, therefore, require locking.
+Locking may be not be needed if multiple independent channels are available.
```
"may not be needed"?

nashif commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+enough to model these. See the *I/O taxonomy* section below.
+
+
+A generic interface
```
Capitalise The Title Please

nashif commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+very low-overhead tracing infrastructure.
+
+
+CTF middle-layer example
```
Capitalise

nashif commented on doc/subsystems/tracing/ctf.rst:
```diff
@@ -0,0 +1,236 @@
+internally and statically at compile-time in the bottom-layer.
+
+
+How to activate?
```
Capitalise

nashif commented on subsys/debug/Kconfig:
```diff
@@ -257,6 +257,15 @@ config TRACING_CPU_STATS_INTERVAL
+config TRACING_CTF_BOTTOM_POSIX
+	bool "CTF backend for the native_posix port, using a file in the host filesystem"
+	depends on TRACING_CTF
+	depends on ARCH_POSIX
```
help missing

nashif commented on subsys/debug/Kconfig:
```diff
@@ -257,6 +257,15 @@ config TRACING_CPU_STATS_INTERVAL
+config TRACING_CTF
+	bool "Tracing via Common Trace Format support"
+	select THREAD_MONITOR
+	select TRACING
```
help missing

MarekPieta commented on subsys/debug/tracing/include/tracing_ctf.h:
```diff
@@ -0,0 +1,39 @@
+}
+#endif
+
+#endif /* _TRACE_CPU_STATS_H */
```
comment should be: _TRACE_CTF_H
